packages/vite/src/node/plugins/define.ts

```diff
@@ -145,36 +135,86 @@ export function definePlugin(config: ResolvedConfig): Plugin {
         return
       }
 
-      const [replacements, pattern] = ssr ? ssrPattern : defaultPattern
+      const [define, pattern] = ssr ? ssrPattern : defaultPattern
+      if (!pattern) return
 
-      if (!pattern) {
-        return null
-      }
+      // Check if our code needs any replacements before running esbuild
+      pattern.lastIndex = 0
+      if (!pattern.test(code)) return
 
-      if (ssr && !isBuild) {
-        // ssr + dev, simple replace
-        return code.replace(pattern, (_, match) => {
-          return '' + replacements[match]
-        })
-      }
+      return await replaceDefine(code, id, define, config)
+    },
+  }
+}
 
-      const s = new MagicString(code)
-      let hasReplaced = false
-      let match: RegExpExecArray | null
+export async function replaceDefine(
+  code: string,
+  id: string,
+  define: Record<string, string>,
+  config: ResolvedConfig,
+): Promise<{ code: string; map: string | null }> {
+  // Because esbuild only allows JSON-serializable values, and `import.meta.env`
+  // may contain values with raw identifiers, making it non-JSON-serializable,
+  // we replace it with a temporary marker and then replace it back after to
+  // workaround it. This means that esbuild is unable to optimize the `import.meta.env`
+  // access, but that's a tradeoff for now.
+  const replacementMarkers: Record<string, string> = {}
+  const env = define['import.meta.env']
+  if (env && !canJsonParse(env)) {
+    const marker = `_${getHash(env, env.length - 2)}_`
+    replacementMarkers[marker] = env
+    define = { ...define, 'import.meta.env': marker }
+  }
 
-      while ((match = pattern.exec(code))) {
-        hasReplaced = true
-        const start = match.index
-        const end = start + match[0].length
-        const replacement = '' + replacements[match[1]]
-        s.update(start, end, replacement)
-      }
+  const result = await transform(code, {
+    loader: 'js',
+    charset: 'utf8',
+    platform: 'neutral',
+    define,
+    sourcefile: id,
+    sourcemap: config.command === 'build' && !!config.build.sourcemap,
+  })
```

Comment:
Shouldn't this `transform` call also be configured by the `esbuild` config option? For example to be able to have a different `charset`. I wonder if the define plugin could be merged with the esbuild plugin (maybe in a different PR to apply the define in the same pass as the minification during build)

Reply:
I'm concerned that inheriting the esbuild options would make it over-transform certain syntax though. Currently the goal is to only replace `define` and nothing else, so it's a bit opinionated.

We could respect some parse-sensitive options like `charset`, but I think the rest should be slowly added. 

Re merging with other plugins, unfortunately that would change the plugin ordering and some post-plugins might expect the `define` already replaced so we can't quite share it.

Reply:
> I'm concerned that inheriting the esbuild options would make it over-transform certain syntax though. Currently the goal is to only replace `define` and nothing else, so it's a bit opinionated.
> 
> We could respect some parse-sensitive options like `charset`, but I think the rest should be slowly added.

Sounds good, I think at least charset is important 👍🏼

> Re merging with other plugins, unfortunately that would change the plugin ordering and some post-plugins might expect the `define` already replaced so we can't quite share it.

I think there may be another breaking change to do in the future and change the plugin ordering. But I understand that this major may not be the moment to do that. Use cases that require the replace to be before post-plugins could add their own. Now that this is done with esbuild, it will be hard to justify in the long term having two passes of esbuild for each file.

Reply:
Yeah I don't think it's great too, but it's kinda the best option we got currently 😅 But luckily we're only running esbuild twice for TS files and contains `import.meta.env`/`define`, so not too bad for now.